mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -45,12 +53,14 @@ def _autolog(log_every_n_epoch=1, log_models=True):
                        are logged after every epoch.
     :param log_models: If ``True``, trained models are logged as MLflow model artifacts.
                        If ``False``, trained models are not logged.
+    :param disable: If ``True``, disables all supported autologging integrations. If ``False``,
+                    enables all supported autologging integrations.
     """
     global every_n_epoch
     every_n_epoch = log_every_n_epoch
 
     def getPLCallback(log_models, metrics_logger):
-        class __MLflowPLCallback(pl.Callback):
+        class __MLflowPLCallback(pl.Callback, metaclass=ExceptionSafeClass):
```

Comment:
This line causes the following error:

```
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass
of the metaclasses of all its bases
```

`pl.Callback` is an abstract class.

Reply:
Found a workaround: https://stackoverflow.com/a/28727066/6943581

Reply:
A workaround:

```python
        class Callback:
            def __getattr__(self, name):
                return pl.Callback().__getattribute__(name)

        class __MLflowPLCallback(Callback, metaclass=ExceptionSafeClass):
```

Reply:
@harupy Is it possible to generate `ExceptionSafeClass` based on a base type? E.g., write a function that provides the metaclass given base type `type` or base type `abc.ABC`?

Reply:
Like this?

```python
def create_ExceptionSafeClass(base):
    class ExceptionSafeClass(base):
        ...

    return ExceptionSafeClass
```

Reply:
Yeah, does that work?

Reply:
For simplicity, assuming there aren’t that many bases used, we can define a couple of these statically. E.g ExceptionSafeClass = function (type) and ExceptionSafeAbstractClass = function(abc.ABC)

Reply:
Tried this

```python
class ExceptionSafeAbstractClass(abc.ABC):
    def __new__(cls, name, bases, dct):
        for m in dct:
            if callable(dct[m]):
                dct[m] = exception_safe_function(dct[m])
        return type.__new__(cls, name, bases + (abc.ABC,), dct)

__MLflowPLCallback(Callback, metaclass=ExceptionSafeAbstractClass)
```

but didn't work.

Reply:
@harupy This seems to work:

```
import abc
class ExceptionSafeAbstractClass(abc.ABCMeta):

    def __new__(mcls, name, bases, namespace, **kwargs):
        for m in namespace:
            if callable(namespace[m]):
                print(namespace[m])
                namespace[m] = exception_safe_function(namespace[m])
        return abc.ABCMeta.__new__(mcls, name, bases, namespace, **kwargs)
```

Can we use this and update the validation logic for `safe_patch` to assert that the metaclass is either `ExceptionSafeClass` or `ExceptionSafeAbstractClass`?

Reply:
@dbczumar 

Should we update the relevent tests?

```diff
 @pytest.mark.usefixtures(test_mode_on.__name__)
-def test_exception_safe_class_exhibits_expected_behavior_in_test_mode():
+@pytest.mark.parametrize(
+    "baseclass,metaclass", [(object, ExceptionSafeClass), (abc.ABC, ExceptionSafeAbstractClass)]
+)
+def test_exception_safe_class_exhibits_expected_behavior_in_test_mode(baseclass, metaclass):
     assert autologging_utils._is_testing()
 
-    class NonThrowingClass(metaclass=ExceptionSafeClass):
+    class NonThrowingClass(baseclass, metaclass=metaclass):
         def function(self):
             return 10
 
@@ -497,7 +504,7 @@ def test_exception_safe_class_exhibits_expected_behavior_in_test_mode():
 
     exc_to_throw = Exception("function error")
 
-    class ThrowingClass(metaclass=ExceptionSafeClass):
+    class ThrowingClass(baseclass, metaclass=metaclass):
         def function(self):
             raise exc_to_throw
```

Reply:
Yes :)

Reply:
done!